routing/checkpoint_predictor.cpp

```diff
@@ -0,0 +1,54 @@
+#include "routing/checkpoint_predictor.hpp"
+
+#include "geometry/mercator.hpp"
+
+#include "base/stl_helpers.hpp"
+
+#include <limits>
+
+namespace routing
+{
+using namespace std;
+
+double CheckpointPredictor::CalculateDeltaMeters(m2::PointD const & from, m2::PointD const & to,
+                                                 m2::PointD const & between) const
+{
+  double const directDist = MercatorBounds::DistanceOnEarth(from, to);
+  double const distThroughPoint = MercatorBounds::DistanceOnEarth(from, between) +
+                                  MercatorBounds::DistanceOnEarth(between, to);
+  return distThroughPoint - directDist;
+}
+
+size_t CheckpointPredictor::PredictPosition(vector<m2::PointD>::const_iterator checkpointsBegin,
```

Comment:
Если передавать сразу вектор, то придется создавать вектор в void RoutingManager::ReorderIntermediatePoints(). Сейчас вызов ForEachIntermediatePoint выглядит так:
```c++
void RoutingManager::ReorderIntermediatePoints()
{
...

  // Note. An added (new) intermediate point is at |positions[0]|. The other points at |positions|
  // is former intermediate points.
  size_t const addingPntOrder = predictor.ForEachIntermediatePoint(
      positions.cbegin() + 1, positions.cend(), positions[0], [&](size_t index, size_t order) {
        CHECK_LESS(index + 1, positions.size(), ());
        // Note. It's necessary to use |index + 1| because it is index in |points|/|positions|
        // but |positions.cbegin() + 1| is passed as a param to CheckpointPredictor::ForEachIntermediatePoint().
        points[index + 1]->SetIntermediateIndex(static_cast<int8_t>(order));
      });
  points[0]->SetIntermediateIndex(static_cast<int8_t>(addingPntOrder));
  routePoints.NotifyChanges();
}
```

Итераторы позволяют не создавать вектор: positions.cbegin() + 1, positions.cend()